src/Wine.cc

```diff
@@ -0,0 +1,175 @@
+/* -*- Mode: C++; tab-width: 8; c-basic-offset: 2; indent-tabs-mode: nil; -*- */
+
+#include <algorithm>
+#include <cstdint>
+#include <list>
+#include <sstream>
+#include <string>
+
+#include "AddressSpace.h"
+#include "Task.h"
+#include "log.h"
+
+using namespace std;
+
+namespace rr {
+
+/* see wine include/winnt.h*/
+#define IMAGE_DOS_SIGNATURE    0x5A4D     /* MZ   */
+#define IMAGE_NT_SIGNATURE     0x00004550 /* PE00 */
+#define IMAGE_NT_OPTIONAL_HDR32_MAGIC      0x10b
+#define IMAGE_NT_OPTIONAL_HDR64_MAGIC      0x20b
+struct __attribute__((packed)) IMAGE_DOS_HEADER {
+  uint16_t e_magic;  /* 00: MZ Header signature */
+  char unused1[0x3a];
+  uint32_t e_lfanew; /* 3c: Offset to extended header */
+};
+struct __attribute__((packed)) IMAGE_FILE_HEADER {
+  char unused1[0x2];
+  uint16_t NumberOfSections;
+  char unused2[0xc];
+  uint16_t SizeOfOptionalHeader;
+  char unused3[0x2];
+};
+struct __attribute__((packed)) IMAGE_OPTIONAL_HEADER32 {
+  uint16_t Magic;
+  char unused1[0xde];
+};
+struct __attribute__((packed)) IMAGE_NT_HEADERS {
+  uint32_t Signature;
+  struct IMAGE_FILE_HEADER FileHeader;
+  struct IMAGE_OPTIONAL_HEADER32 OptionalHeader;
+};
+struct __attribute__((packed)) IMAGE_SECTION_HEADER {
+  char unused1[0xc];
+  uint32_t VirtualAddress;
+  char unused2[0x18];
+};
+#define IMAGE_FIRST_SECTION(ntheader) \
+  ((struct IMAGE_SECTION_HEADER*)((char *)&(ntheader)->OptionalHeader + \
+                           (ntheader)->FileHeader.SizeOfOptionalHeader))
+
+
+string create_library_response_from_mappings(Task *t) {
+
+  // create an own list of the current mappings
+  list<KernelMapping> mapping_list;
+  for (KernelMapIterator it(t); !it.at_end(); ++it) {
+    auto km = it.current();
+    if (km.fsname().substr(0, 7) == "/memfd:")
+      continue;
+    if (km.is_vdso() || km.is_heap() || km.is_stack() || km.is_vvar() || km.is_vsyscall())
+      continue;
+    LOG(debug) << "Possible mapping for GDB reply: " << km;
+    mapping_list.push_back(km);
+  }
+
+  struct library {
+    library() : has_exec(false) {};
+    vector<remote_ptr<void>> segments;
+    bool has_exec;
+  };
+  map<string, library> libraries;
+
+  // iterate through the mappings and search for PE libraries
+  // see wine, packet_query_libraries_cb
+  char buf[0x400];
+  for (auto it = mapping_list.begin(); it != mapping_list.end(); it++) {
+    auto km = *it;
+    bool ok = true;
+    t->read_bytes_helper(km.start(), sizeof(buf), buf, &ok);
+    if (ok) {
+      struct IMAGE_DOS_HEADER* dosheader = (struct IMAGE_DOS_HEADER*)buf;
+      struct IMAGE_NT_HEADERS* ntheader = (struct IMAGE_NT_HEADERS*)(buf + dosheader->e_lfanew);
```

Comment:
Technically this is undefined behavior. Don't construct the pointer until you've tested that there is room for `IMAGE_NT_HEADERS` in `buf`.